Get count param

tazwar22 opened the pull request:
Addressing #50 

Current Implementation:
- Simply returns the first "count" (1,2,3... N) articles (specified in route parameter)
- Not robust to count > collection size
- 1 route Handler for both queried/non-queried versions of GET Article 

Open to suggestions for better design.

tazwar22 committed:
Limiting number of articles returned

tazwar22 committed:
Added vanilla implementation; simply returns first-COUNT articles

tarikeshaq commented on backend/routes.go:
```diff
@@ -25,7 +26,9 @@ func (s *server) handleRoutes() error {
-	router.HandleFunc("/articles", s.handleGetArticles()).Methods("GET")
+	//Adding route param...
+	router.HandleFunc("/articles/", s.handleGetArticles()).Methods("GET")
+	router.HandleFunc("/articles/{count}", s.handleGetArticles()).Methods("GET")
```
Sweet! I wonder if we can use optional query parameters instead of a path.. i.e something like this https://stackoverflow.com/questions/43379942/how-to-have-an-optional-query-in-get-request-using-gorilla-mux

I know I said to use `vars` in the issue, but this seems a little neater since they are optional (and otherwise we would return all articles if it's not provided)

That way, you can have something like this in the handler,
```
v := r.URL.Query()
count := v.Get("count")
// ... We can add the strconv
```

And we would avoid setup an extra `handleFunc` up here

tarikeshaq commented on backend/routes.go:
```diff
@@ -93,18 +98,34 @@ func (s *server) handleGetArticles() http.HandlerFunc {
+			fmt.Printf("Limiting number of articles to  %d \n", numArticles)
+		}
+
+		for _, a := range results[:numArticles] {
```
We spoke about it in slack - let's try to do the counting in the database instead of loading all the values into memory first!

tarikeshaq commented on backend/routes.go:
```diff
@@ -93,18 +98,34 @@ func (s *server) handleGetArticles() http.HandlerFunc {
+		
+		//Get number of articles necessary
+		//ASSUMPTION: count <= SIZE(RESULTS); will fix this later
+		numArticles, err := strconv.Atoi(mux.Vars(req)["count"])
```
The linter is complaining because we aren't checking the error value, you probably already have that in mind, just wanted to clarify what the liner is complaining about 😄

tarikeshaq reviewed (commented):
Good work @tazwar22 ! Dropped a couple of comments, if you'd like me to comment on anything specific, let me know

tazwar22 committed:
Updating DB Interface to allow retrieval of subsets; yet to implement

tazwar22 committed:
Updated mongo.go function implementation to allow returning limited articles

tazwar22 committed:
Updated variable name to something more intuitive

tazwar22 commented:
@tarikeshaq 
Updated implementation:
1.  Followed suggestion for using **Optional** Query parameters; can extend this logic to other routes in the future as need be
2. ` mongo.db` Article Retrieval function only collects the first N-articles
3. Deals with count > N appropriately

There's probably a better way to Iterate over a limited subset of the Collection, compared to the very simple way in which I implemented it, so there's significant room for refactoring the function! Open to an alternative suggestion.

tarikeshaq commented on backend/db/db.go:
```diff
@@ -63,6 +63,8 @@ func (c *Database) CreateNewArticle(url string, title string, source string) (st
-	return c.FindAllArticles()
+// numArticles: number of Articles to return from Collection,
+// NOTE:  If -1 is passed in, we return ALL Articles
+func (c *Database) GetAllArticles(numArticles int) ([]Article, error) {
```
Hmm can we define a seperate function for this?

The name `GetAllArticles` suggests we are getting all articles. Lets define a separate function that takes in a parameter. We can do some refactoring in `mongo.go` to help us with the common code

This way we don't have to carry a special behavior for the value of `-1`

tarikeshaq commented on backend/db/mongo.go:
```diff
@@ -41,8 +40,11 @@ func (c *Database) insertArticle(article Article) (string, error) {
-func (c *Database) FindAllArticles() ([]Article, error) {
+
+
+// fetch ALL Article from mongo if numArticles == -1 (FLAG)
```
I appreciate the comments! Let's clean them up (we don't need the unnecessary capitalization, and it can be in plain English) once we've refactored :)

Hopefully, we can also later document all our exported functions this way for go doc :)

tarikeshaq commented on backend/db/mongo.go:
```diff
@@ -57,17 +59,28 @@ func (c *Database) FindAllArticles() ([]Article, error) {
 	var articles []Article
 
+	numObtained := 0
 	for cursor.Next(context.TODO()) {
```
I would prefer if we have a loop where we put numObtained in the condition, saves us from the extra if statement. And assuming we have this in a separate function to get rid of the -1 special handling all will be good
```go
for cursor.Next(context.TODO()) && numObtained < numArticles {
.
.
}

```

tarikeshaq commented on backend/db/mongo.go:
```diff
@@ -57,17 +59,28 @@ func (c *Database) FindAllArticles() ([]Article, error) {
-
 }
 
+
```
Lets get rid of this extra spacing. One empty line is ok

tarikeshaq commented on backend/routes.go:
```diff
@@ -25,7 +26,9 @@ func (s *server) handleRoutes() error {
-	router.HandleFunc("/articles", s.handleGetArticles()).Methods("GET")
+	//Adding route param...
+	router.HandleFunc("/articles/", s.handleGetArticles()).Methods("GET")
+	// router.HandleFunc("/articles/{count}", s.handleGetArticles()).Methods("GET")
```
We can get rid of those comments
```suggestion
```

tarikeshaq commented on backend/routes.go:
```diff
@@ -78,21 +81,33 @@ func (s *server) handleAddUser() http.HandlerFunc {
 //GET: endpoint for all articles
 func (s *server) handleGetArticles() http.HandlerFunc {
 	return func(w http.ResponseWriter, req *http.Request) {
+
```
Empty lines are unnecessary at the top. At most, we should have one empty line between every logically independent piece of code. Smaller functions are easier to look at :)

tarikeshaq commented on backend/routes.go:
```diff
@@ -78,21 +81,33 @@ func (s *server) handleAddUser() http.HandlerFunc {
+			numArticles, _ = strconv.Atoi(v.Get("count"))
+		}
+
+		//IF 
```
```suggestion
```

tarikeshaq commented on backend/routes.go:
```diff
@@ -78,21 +81,33 @@ func (s *server) handleAddUser() http.HandlerFunc {
 
+		//No count value? RETURN ALL
+		if v.Get("count") != "" {
+			numArticles, _ = strconv.Atoi(v.Get("count"))
```
Lets not discard the error. It could be a valid issue with the caller sending an invalid count.

We should return an error to the caller in this case with `http.StatusBadRequest`

tarikeshaq commented on backend/routes.go:
```diff
@@ -78,21 +81,33 @@ func (s *server) handleAddUser() http.HandlerFunc {
 
 		var articles = make([]ArticleJson, 0)
 		var article ArticleJson
+
```
```suggestion
```

tarikeshaq commented on backend/routes.go:
```diff
@@ -101,10 +116,25 @@ func (s *server) handleGetArticles() http.HandlerFunc {
 	}
 }
 
+
```
Ditto on too many unnecessary empty lines

tarikeshaq commented on backend/db/db.go:
```diff
@@ -63,6 +63,8 @@ func (c *Database) CreateNewArticle(url string, title string, source string) (st
+// numArticles: number of Articles to return from Collection,
+// NOTE:  If -1 is passed in, we return ALL Articles
+func (c *Database) GetAllArticles(numArticles int) ([]Article, error) {
+		return c.FindAllArticles(numArticles)
```
Too much indentation here

tarikeshaq reviewed (commented):
Sweet work @tazwar22 I really like where this is heading 🔥 💯 

I dropped a couple of comments, but keep going!

A few of them can be solved by running `go fmt` before committing your code, `go fmt` will format all the go code in the directory it is run so that it is by go standards! (getting rid of extra white spaces etc)

In your case, since you have changes on both the `backend` directory and the `db` directory, run it in both - it will automagically format the code!
